okio/src/appleMain/kotlin/okio/Buffer.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import kotlinx.cinterop.ByteVarOf
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.plus
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.usePinned
+import platform.Foundation.NSData
+import platform.Foundation.NSMutableData
+import platform.Foundation.create
+import platform.darwin.NSInteger
+import platform.posix.memcpy
+
+fun Buffer.write(
+  source: NSData,
+  offset: Int = 0,
+  byteCount: Int = source.length.toInt()
+): Buffer = apply {
+  var offset = offset
+  checkOffsetAndCount(source.length.toLong(), offset.toLong(), byteCount.toLong())
+
+  val limit = offset + byteCount
+  while (offset < limit) {
+    val tail = writableSegment(1)
+
+    val toCopy = minOf(limit - offset, Segment.SIZE - tail.limit)
+    tail.data.usePinned { pinned ->
+      memcpy(
+        __dst = pinned.addressOf(tail.limit),
+        __src = (source.bytes as CPointer<ByteVarOf<*>>) + offset,
+        __n = toCopy.toULong()
+      )
+    }
+
+    offset += toCopy
+    tail.limit += toCopy
+  }
+
+  size += byteCount.toLong()
+}
+
+fun Buffer.readNSData(byteCount: NSInteger = size): NSData {
+  require(byteCount >= 0 && byteCount <= Int.MAX_VALUE) { "byteCount: $byteCount" }
+  if (size < byteCount) throw EOFException()
+
+  val length = byteCount.toULong()
```

Comment:
should the parameter be a ULong, if that’s NSData’s length type?

Reply:
Good call regarding types. I copied method signatures from `commonMain/Buffer`, but it does make sense to use types that `NSData` uses.

In general, Foundation numeric types are typealiased to Kotlin types, e.g.:

```kotlin
package platform.darwin

public typealias NSInteger = kotlin.Long
public typealias NSUInteger = kotlin.ULong
```

So when this code is called from Kotlin there should be no difference in whether `NSInteger` or `Long` was in the method signature. However, using darwin types should make Swift/ObjC call sites more ergonomic.

Reply:
afaff8f